File: data/src/main/java/com/neuifo/data/domain/utils/ClassConstructorLoader.java
package com.neuifo.data.domain.utils;

import java.lang.reflect.Constructor;
import java.util.Arrays;

/**
 * Created by neuifo on 2016/11/16.
 */

public class ClassConstructorLoader {

    public static <C> Constructor<C> getAppropriateConstructor(Class<C> c, Object[] initArgs) {
        if (initArgs == null)
            initArgs = new Object[0];
        for (Constructor con : c.getDeclaredConstructors()) {
            Class[] types = con.getParameterTypes();
            if (types.length != initArgs.length)
                continue;
            boolean match = true;
            for (int i = 0; i < types.length; i++) {
                Class<?> need = types[i], got = initArgs[i].getClass();
                if (!need.isAssignableFrom(got)) {
                    if (need.isPrimitive()) {
                        match = (int.class.equals(need) && Integer.class.equals(got))
                                || (long.class.equals(need) && Long.class.equals(got))
                                || (char.class.equals(need) && Character.class.equals(got))
                                || (short.class.equals(need) && Short.class.equals(got))
                                || (boolean.class.equals(need) && Boolean.class.equals(got))
                                || (byte.class.equals(need) && Byte.class.equals(got)
                                || (float.class.equals(need) && Float.class.equals(got)));
                        // if (!match)System.out.println("Not match " + need + "; " + got + "; " + initArgs[i]);
                    } else {
                        // System.out.println("Not match " + need + "; " + got + "; " + initArgs[i]);
                        match = false;
                    }
                }
                if (!match)
                    break;
            }
            if (match)
                return con;
        }
        throw new IllegalArgumentException("Cannot find an appropriate constructor for class " + c + " and arguments " + Arrays.toString(initArgs));
    }
}

File: data/src/main/java/com/neuifo/data/domain/utils/ClassConstructorLoader.kt
package com.neuifo.data.domain.utils

import java.lang.reflect.Constructor
import java.util.*

/**
 * Created by neuifo on 2016/11/16.
 */
object ClassConstructorLoader {
    fun <C> getAppropriateConstructor(
        c: Class<C>,
        initArgs: Array<Any?>?
    ): Constructor<C> {
        var initArgs = initArgs
        if (initArgs == null) initArgs = arrayOfNulls(0)
        for (con in c.declaredConstructors) {
            val types = con.parameterTypes
            if (types.size != initArgs.size) continue
            var match = true
            for (i in types.indices) {
                val need = types[i]
                val got: Class<*> = initArgs[i]!!.javaClass
                if (!need.isAssignableFrom(got)) {
                    match = if (need.isPrimitive) {
                        (Int::class.javaPrimitiveType == need && Int::class.java == got
                                || Long::class.javaPrimitiveType == need && Long::class.java == got
                                || Char::class.javaPrimitiveType == need && Char::class.java == got
                                || Short::class.javaPrimitiveType == need && Short::class.java == got
                                || Boolean::class.javaPrimitiveType == need && Boolean::class.java == got
                                || (Byte::class.javaPrimitiveType == need && Byte::class.java == got
                                || Float::class.javaPrimitiveType == need && Float::class.java == got))
                        // if (!match)System.out.println("Not match " + need + "; " + got + "; " + initArgs[i]);
                    } else {
                        // System.out.println("Not match " + need + "; " + got + "; " + initArgs[i]);
                        false
                    }
                }
                if (!match) break
            }
            if (match) return con as Constructor<C>
        }
        throw IllegalArgumentException(
            "Cannot find an appropriate constructor for class $c and arguments " + Arrays.toString(
                initArgs
            )
        )
    }
}